src/renderer/atlas/dwrite.cpp

```diff
@@ -0,0 +1,169 @@
+    static constexpr auto norm24 = static_cast<float>(static_cast<double>(0x100) / (255) * 4);
+
+#pragma warning(suppress : 26451) // Arithmetic overflow: Using operator '+' on a 4 byte value and then casting the result to a 8 byte value.
+    const auto index = clamp<ptrdiff_t>(static_cast<ptrdiff_t>(gamma * 10.0f + 0.5f), 10, 22) - 10;
+#pragma warning(suppress : 26446) // Prefer to use gsl::at() instead of unchecked subscript operator (bounds.4).
+#pragma warning(suppress : 26482) // Only index into arrays using constant expressions (bounds.2).
+    const auto& ratios = gammaIncorrectTargetRatios[index];
+
+    out[0] = norm13 * ratios[0];
+    out[1] = norm24 * ratios[1];
+    out[2] = norm13 * ratios[2];
+    out[3] = norm24 * ratios[3];
+}
+
+// This belongs to isThinFontFamily().
+// Keep this in alphabetical order, or the loop will break.
+// Keep thinFontFamilyNamesMaxWithNull updated.
+static constexpr std::array thinFontFamilyNames{
+    L"Courier New",
+    L"Fixed Miriam Transparent",
+    L"Miriam Fixed",
+    L"Rod",
+    L"Rod Transparent",
+    L"Simplified Arabic Fixed"
+};
+static constexpr size_t thinFontFamilyNamesMaxLengthWithNull = 25;
+
+// DWrite_IsThinFontFamily returns true if the specified family name is in our hard-coded list of "thin fonts".
+// These are fonts that require special rendering because their strokes are too thin.
+//
+// The history of these fonts is interesting. The glyph outlines were originally created by digitizing the typeballs of
+// IBM Selectric typewriters. Digitizing the metal typeballs yielded very precise outlines. However, the strokes are
+// consistently too thin in comparison with the corresponding typewritten characters because the thickness of the
+// typewriter ribbon was not accounted for. This didn't matter in the earliest versions of Windows because the screen
+// resolution was not that high and you could not have a stroke thinner than one pixel. However, with the introduction
+// of anti-aliasing the thin strokes manifested in text that was too light. By this time, it was too late to change
+// the fonts so instead a special case was added to render these fonts differently.
+//
+// ---
+//
+// The canonical family name is a font's family English name, when
+// * There's a corresponding font face name with the same language ID
+// * If multiple such pairs exist, en-us is preferred
+// * Otherwise (if en-us is not a translation) it's the lowest LCID
+//
+// However my (lhecker) understanding is that none of the thinFontFamilyNames come without an en-us translation.
+// As such you can simply get the en-us name of the font from a IDWriteFontCollection for instance.
+// See the overloaded alternative version of isThinFontFamily.
+bool DWrite_IsThinFontFamily(const wchar_t* canonicalFamilyName) noexcept
+{
+    int n = 0;
+
+    for (const auto familyName : thinFontFamilyNames)
+    {
+        n = wcscmp(canonicalFamilyName, familyName);
+        if (n <= 0)

```

Comment:
The trickery of the negative here isn't immediately comprehensible to me. Explain (in a comment)?

Reply:
`wcscmp`, just like other cmp functions (`memcmp`), returns `-1` if the first argument has a character of smaller numeric value than the second argument. `wcscmp(L"foo", L"foz")` returns `-1`.
Due to this reason, this loop acts like a call to `std::lower_bound`, but with linear search, because binary search is kinda overkill for just a few values (log2(x) is close to linear for small x, resulting in 2.6 comparisons with binary search and 3 comparisons with linear search on average here).

I can comment it but I think a long comment would be boring, since the loop cannot be meaningfully improved, unless it's entirely rewritten from scratch.

Reply:
Right... I added a comment explaining the "why" instead of the "how". That seems the most helpful to me. 🙂